### ISSUE TITLE
Backend process not properly terminated

### ISSUE THREAD

**Issue by spakka** [Exa-Networks/exabgp#548]
Hi Thomas,

One of our backend processes that ExaBGP feeds BGP messages to, went a bit wrong. ExaBGP tried to respawn it, and eventually gave up:
```
exabgp.log-20161120:Nov 19 00:00:52 rrc19 exabgp: 5388   processes     Issue while sending data to our helper program, it left us restarting it
exabgp.log-20161120:Nov 19 00:00:52 rrc19 exabgp: 5388   processes     Terminating process receive-routes
exabgp.log-20161120:Nov 19 00:00:52 rrc19 exabgp: 5388   processes     Forked process receive-routes
exabgp.log-20161120:Nov 19 00:01:02 rrc19 exabgp: 5388   processes     Issue while sending data to our helper program, it left us restarting it
exabgp.log-20161120:Nov 19 00:01:03 rrc19 exabgp: 5388   processes     Terminating process receive-routes
exabgp.log-20161120:Nov 19 00:01:03 rrc19 exabgp: 5388   processes     Forked process receive-routes
exabgp.log-20161120:Nov 19 00:01:08 rrc19 exabgp: 5388   processes     Issue while sending data to our helper program, it left us restarting it
exabgp.log-20161120:Nov 19 00:01:08 rrc19 exabgp: 5388   processes     Terminating process receive-routes
exabgp.log-20161120:Nov 19 00:01:09 rrc19 exabgp: 5388   processes     Forked process receive-routes
exabgp.log-20161120:Nov 19 00:01:13 rrc19 exabgp: 5388   processes     Issue while sending data to our helper program, it left us restarting it
exabgp.log-20161120:Nov 19 00:01:13 rrc19 exabgp: 5388   processes     Terminating process receive-routes
exabgp.log-20161120:Nov 19 00:01:13 rrc19 exabgp: 5388   processes     Forked process receive-routes
exabgp.log-20161120:Nov 19 00:01:20 rrc19 exabgp: 5388   processes     Issue while sending data to our helper program, it left us restarting it
exabgp.log-20161120:Nov 19 00:01:20 rrc19 exabgp: 5388   processes     Terminating process receive-routes
exabgp.log-20161120:Nov 19 00:01:20 rrc19 exabgp: 5388   processes     Forked process receive-routes
exabgp.log-20161120:Nov 19 00:01:26 rrc19 exabgp: 5388   processes     Issue while sending data to our helper program, it left us restarting it
exabgp.log-20161120:Nov 19 00:01:26 rrc19 exabgp: 5388   processes     Terminating process receive-routes
exabgp.log-20161120:Nov 19 00:01:27 rrc19 exabgp: 5388   processes     Forked process receive-routes
exabgp.log-20161120:Nov 19 00:01:34 rrc19 exabgp: 5388   processes     Issue while sending data to our helper program, it left us restarting it
exabgp.log-20161120:Nov 19 00:01:34 rrc19 exabgp: 5388   processes     Terminating process receive-routes
exabgp.log-20161120:Nov 19 00:01:34 rrc19 exabgp: 5388   processes     Forked process receive-routes
exabgp.log-20161120:Nov 19 00:01:38 rrc19 exabgp: 5388   processes     Issue while sending data to our helper program, it left us restarting it
exabgp.log-20161120:Nov 19 00:01:38 rrc19 exabgp: 5388   processes     Terminating process receive-routes
exabgp.log-20161120:Nov 19 00:01:38 rrc19 exabgp: 5388   processes     Forked process receive-routes
exabgp.log-20161120:Nov 19 00:01:38 rrc19 exabgp: 5388   processes     Too many respawn for receive-routes (5) terminating program
```
However, the process was left running. Except it was not being fed any data by ExaBGP. This meant our monitoring did not detect that the process was down.

The backend process kept running, its STDIN was still hooked to the ExaBGP process, but I guess ExaBGP took it out of the list of processes it should send data to. The process stopped crashing, btw, because it was only crashing when it was sent data, and that stopped happening :-)

We've recently found that this has happened more than once, for different instances of ExaBGP.
This is running ExaBGP 3.4.15

If you need any more info, just let me know.

Cheers,
Colin

**Comment by thomas-mangin** [comment]
Hello Colin, 

Do you have a faulty process I can use - to see what happens.
Also more logs (to see when in the code the exception was caught) would be good.

It does not look my code try to find PID and kill the process when something goes wrong but I do not have the time to investigate right now.

You can try this patch but it is really a "trying to guess one"
```
diff --git a/lib/exabgp/reactor/api/processes.py b/lib/exabgp/reactor/api/processes.py
index 94d305bb..93ec5eca 100644
--- a/lib/exabgp/reactor/api/processes.py
+++ b/lib/exabgp/reactor/api/processes.py
@@ -126,6 +126,7 @@ class Processes (object):
                                                                "Too many respawn for %s (%d) terminating program" % (process,self.respawn_number),
                                                                'critical'
                                                        )
+                                                       self._terminate(process)
                                                        raise ProcessError()
                                        else:
                                                # reset long time since last respawn
@@ -135,6 +136,7 @@ class Processes (object):
                                        self._respawning[process] = {around_now: 1}

                except (subprocess.CalledProcessError,OSError,ValueError),exc:
+                       self._terminate(process)
                        self._broken.append(process)
                        self.logger.processes("Could not start process %s" % process)
                        self.logger.processes("reason: %s" % str(exc))
@@ -218,8 +220,10 @@ class Processes (object):
                                self._process[process].stdin.write('%s\n' % string)
                        except IOError,exc:
                                self._broken.append(process)
+                               self._terminate(process)
                                if exc.errno == errno.EPIPE:
                                        self._broken.append(process)
+                                       self._terminate(process)
                                        self.logger.processes("Issue while sending data to our helper program")
                                        raise ProcessError()
                                else:
```

**Comment by spakka** [comment]
Let me knock something together and see if I can replicate it.
Will let you know shortly

**Comment by spakka** [comment]
I've managed to replicate it, as well as #549 - have emailed you all the details. Let me know if there's anything else you need!

**Comment by thomas-mangin** [comment]
@spakka - Can not seems to find your mail :-(

**Comment by spakka** [comment]
Re-sent

**Comment by thomas-mangin** [comment]
fixed in recent ExaBGP see #549 for details.
